qiskit_ibm_runtime/api/session.py

```diff
@@ -264,6 +270,30 @@ def request(  # type: ignore[override]
         headers = self.headers.copy()
         headers.update(kwargs.pop("headers", {}))
 
+        # Set default caller
+        headers.update({"X-Qx-Client-Application": f"{CLIENT_APPLICATION}/qiskit"})
+
+        # Use PurePath in order to support arbitrary path formats
+        callers = {PurePath("qiskit/"), "qiskit_"}
+
+        for frame in inspect.stack():
+            frame_path = str(PurePath(frame.filename))
+            for caller in callers:
+                if str(caller) in frame_path:
+                    caller_str = str(caller) + frame_path.split(str(caller), 1)[-1]
+                    sanitized_caller_str = caller_str.replace("/", "~")
```

Comment:
nit - can we use `.` instead of `~`? After looking at your example, `qiskit_nature.second_q.algorithms.ground_state_solvers.ground_state_eigensolver.py` would looks a lot nicer lol 

Also I think we can break out of both for loops once a matching str is found?

Reply:
Using `.` instead of `~` would create "collisions" in every single path due to file extensions including a `.` e.g. `ground_state_eigensolver.py`. It would be unnecessarily complicated to replace all the "right" `.` with `/` again when e.g. resubstituting the characters in the backend. 
Although we  could do this a split called on the string would create weird segments as well. One might expect that you get all the path segments when you call split with our chosen delimiter but with a `.`, you would get the `py` as an unexpected segment as well.

I don't think that we can break out of the loops. We are iterating through the stacktrace which consists out of many calls in `qiskit-ibm-runtime` and many of them also match the criteria. We want the last matching one since that one should be the one that is closest to the user.

Reply:
> when e.g. resubstituting the characters in the backend.

I don't think we are doing any resubstitution in the backend? I think it's just stored in the DB and then be queried. But I don't feel too strongly if you don't want to change it.

> We want the last matching one since that one should be the one that is closest to the user.

But isn't the list returned by `inspect.stack()` already ordered by closest to the user first?

Reply:
So far, we are not doing any resubstitution but I'd love to at least have the option which is why I think I will leave this part as is. 

If we take the first one of the stack, we would always get `qiskit-ibm-runtime/qiskit_ibm_runtime/api/session.py` which isn't the result we want.

Reply:
Good point 😄 The latest commit implemented this change request.